filesystem/File.php

```diff
@@ -504,103 +545,67 @@ public function onAfterUpload() {
-	 * so it has to handle both files and folders.
+	 * This method will update the File {@see DBFile} field value on success, so it must be called
+	 * before writing to the database
 	 *
-	 * Assumes that the "Filename" property was previously updated, either directly or indirectly.
-	 * (it might have been influenced by {@link setName()} or {@link setParentID()} before).
+	 * @param bool True if changed
 	 */
 	public function updateFilesystem() {
-		if(!$this->config()->update_filesystem) return false;
-
-		// Regenerate "Filename", just to be sure
-		$this->setField('Filename', $this->getRelativePath());
-
-		// If certain elements are changed, update the filesystem reference
-		if(!$this->isChanged('Filename')) return false;
-
-		$changedFields = $this->getChangedFields();
-		$pathBefore = $changedFields['Filename']['before'];
-		$pathAfter = $changedFields['Filename']['after'];
-
-		// If the file or folder didn't exist before, don't rename - its created
-		if(!$pathBefore) return;
-
-		$pathBeforeAbs = Director::getAbsFile($pathBefore);
-		$pathAfterAbs = Director::getAbsFile($pathAfter);
-
-		// TODO Fix Filetest->testCreateWithFilenameWithSubfolder() to enable this
-		// // Create parent folders recursively in database and filesystem
-		// if(!is_a($this, 'Folder')) {
-		// 	$folder = Folder::findOrMake(dirname($pathAfterAbs));
-		// 	if($folder) $this->ParentID = $folder->ID;
-		// }
-
-		// Check that original file or folder exists, and rename on filesystem if required.
-		// The folder of the path might've already been renamed by Folder->updateFilesystem()
-		// before any filesystem update on contained file or subfolder records is triggered.
-		if(!file_exists($pathAfterAbs)) {
-			if(!is_a($this, 'Folder')) {
-				// Only throw a fatal error if *both* before and after paths don't exist.
-				if(!file_exists($pathBeforeAbs)) {
-					throw new Exception("Cannot move $pathBeforeAbs to $pathAfterAbs - $pathBeforeAbs doesn't exist");
-				}
+		if(!$this->config()->update_filesystem) {
+			return false;
+		}
 
-				// Check that target directory (not the file itself) exists.
-				// Only check if we're dealing with a file, otherwise the folder will need to be created
-				if(!file_exists(dirname($pathAfterAbs))) {
-					throw new Exception("Cannot move $pathBeforeAbs to $pathAfterAbs - Directory " . dirname($pathAfter)
-						. " doesn't exist");
-				}
-			}
+		// Check the file exists
+		if(!$this->File->exists()) {
+			return false;
+		}
 
-			// Rename file or folder
-			$success = rename($pathBeforeAbs, $pathAfterAbs);
-			if(!$success) throw new Exception("Cannot move $pathBeforeAbs to $pathAfterAbs");
+		// Check path updated record will point to
+		// If no changes necessary, skip
+		$pathBefore = $this->File->getFilename();
+		$pathAfter = $this->getFilename();
+		if($pathAfter === $pathBefore) {
+			return false;
 		}
 
+		// Copy record to new location via stream
```

Comment:
This changes the semantics of `updateFilesystem()` to *copy* the file, rather than *move* it. What's the reason for keeping it in its old place? Also, reading from a stream while writing to it sounds like a recipe for edge cases in some adapters. Flysystem has `rename()`?